Add com.flashforge.FlashPrint.

Yxogenium opened the pull request:
FlashPrint is the slicer for use with Flashforge printers: http://www.flashforge.com/flashprint/.

It is not redistributable, therefore I followed the approach found in the manifest for com.spotify.Client.
The buttons sometime flicker but all the features I tested seemed to work as expected.

One oddity though: when run in flatpak, FlashPrint complains unless `"--device=dri"` is added to the `finish-args`:
```
libGL error: MESA-LOADER: failed to retrieve device information
libGL error: Version 4 or later of flush extension not found
libGL error: failed to load driver: i915
```
However, if this flag is added to the manifest, some buttons no longer show up in FlashPrint, even though there is no error message. This is why I decided to omit this flag.

nedrichards commented:
bot, build com.flashforge.FlashPrint
testbot, build com.flashforge.FlashPrint

nedrichards commented on com.flashforge.FlashPrint.appdata.xml:
```diff
@@ -0,0 +1,25 @@
+    <p>FlashPrint is a beginner-friendly slicer developped in-house by Flashforge for use
+      with its line of 3D printers, featuring a 3D viewer, scaling and cutting abilities,
+      and multiple printer related configuration options. It also sports an expert mode for
+      more experienced users.</p>
```
Can you put something like:
> NOTE: This wrapper is not verified by, affiliated with, or supported by FlashForge

here? It's important to make sure that users know that this isn't officially distributed for these sort of apps..

scx commented on com.flashforge.FlashPrint.json:
```diff
@@ -0,0 +1,78 @@
+  ],
+  "finish-args": [
+    "--filesystem=home",
+    "--socket=x11"
```
For X11, you need the `--share=ipc` as well.
And if this application uses OpenGL acceleration, you should also put the `--device=dri` here.

scx commented on com.flashforge.FlashPrint.appdata.xml:
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
```
The AppData file lacks some required tags, such as `launchable` tag or `screenshot` tag.
https://github.com/flathub/flathub/wiki/AppData-Guidelines

Yxogenium replied:
Should I commit the screenshots in the repo ?

The .desktop is pulled from the .deb at install time. Can I link to it in the AppData file ?

Yxogenium replied:
OK, I think it's done the way you recommended now.

nedrichards commented:
The app requiring but being buggy under GPU acceleration sounds like something that should be reported to and fixed upstream.

Have you spoken to FlashForge to see if they're interested in helping maintain this app?

Yxogenium commented:
> The app requiring but being buggy under GPU acceleration sounds like something that should be reported to and fixed upstream.
> 
> Have you spoken to FlashForge to see if they're interested in helping maintain this app?

I did send an email to support@ff3dp.com, but I have not heard back yet.

I have investigated the issue further : it seems to be linked to wayland:
* if I run FlashPrint under xorg (either with `env LD_LIBRARY_PATH="/opt/qt595/lib/" ./FlashPrint`, or in a Flatpak with or without dri) it works flawlessly.
* if I run FlashPrint under Wayland with `env LD_LIBRARY_PATH="/opt/qt595/lib/" ./FlashPrint` or in a Flatpak with dri, the buttons do not show up
* if I run FlashPrint in a Flatpak without dri, the buttons are janky but do show up.

Yxogenium commented:
In my most recent commit, I added back `--device=dri` (mostly as a test): Flashprint works correctly under Xorg but not under Wayland.

I added `ipc` as suggested by scx. I also granted access to the wayland socket as it seems to solve some Qt issues related to theming.

After this changes, FlashPrint only complains that it has no network access (apparently it only needs network access to check for firmware updates, but since it needs udev rules to communicate with the printer, I deemed it unnecessary to give it network access).

nedrichards commented:
bot, build com.flashforge.FlashPrint
testbot, build com.flashforge.FlashPrint

Yxogenium commented:
They answered ! But it seems that we'll have to wait some more before we get in touch with the developers:
> Hello ,
> Thank you very much for the message, this is [] from Flashforge and I will be your support contact.
> I will submit your email to our software R&D team and hope them can make improvements in the future.
> And they are on Chinese New Year Holiday now,they will come back on Feb 14th,2019 then I can communicate with them.

barthalion commented:
> Flashprint works correctly under Xorg but not under Wayland.

Not under Wayland directly or Xwayland?

Yxogenium commented on com.flashforge.FlashPrint.json:
```diff
@@ -0,0 +1,81 @@
+{
+  "app-id": "com.flashforge.FlashPrint",
+  "runtime": "org.kde.Platform",
+  "runtime-version": "5.9",
```
Yes, FlashPrint only runs with Qt versions up to 5.9. I have notified the developers in my email, and the issue pops up in google groups dating back January 2018 so they are probably already aware of this shortcoming.

Yxogenium commented:
> > Flashprint works correctly under Xorg but not under Wayland.
> 
> Not under Wayland directly or Xwayland?

XWayland I guess, it's even worse when I force Wayland:
```console
$ QT_QPA_PLATFORM=wayland flatpak run com.flashforge.FlashPrint
QSocketNotifier: Can only be used with threads started with QThread
/app/bin/FlashPrint: line 3:     3 Aborted                 (core dumped) /app/extra/share/FlashPrint/FlashPrint "$@"
```

Yxogenium commented on com.flashforge.FlashPrint.appdata.xml:
```diff
@@ -0,0 +1,26 @@
+      with its line of 3D printers, featuring a 3D viewer, scaling and cutting abilities,
+      and multiple printer related configuration options. It also sports an expert mode for
+      more experienced users.
+      NOTE: This wrapper is not verified by, affiliated with, or supported by FlashForge.</p>
```
Done.

Yxogenium commented on com.flashforge.FlashPrint.appdata.xml:
```diff
@@ -0,0 +1,26 @@
+    <category>Development</category>
+    <category>Science</category>
+  </categories>
+  <icon type="remote" height="72" width="72">https://www.flashforge.com/wp-content/uploads/2016/06/favicon_72x72.png</icon>
```
Updated.

Yxogenium commented on flashprint-bin:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+/app/extra/share/FlashPrint/FlashPrint "$@"
```
Done.

nedrichards commented:
bot, build com.flashforge.FlashPrint
testbot, build com.flashforge.FlashPrint

barthalion commented on com.flashforge.FlashPrint.json:
```diff
@@ -0,0 +1,81 @@
+  ],
+  "finish-args": [
+    "--filesystem=home",
+    "--socket=wayland",
```
I guess it shouldn't specify wayland here if it's broken. `x11` should be enough to use xwayland

CommunicationAnimale replied:
You are right, the flag does not seem to have any effect. I will remove it.

Yxogenium replied:
Indeed, I have now removed this flag.

barthalion commented on com.flashforge.FlashPrint.json:
```diff
@@ -0,0 +1,81 @@
+    },
+    "shared-modules/glu/glu-9.0.0.json",
+    {
+      "name": "libudev1",
```
We also have udev in shared-modules but I'm not sure about provided soname version.

Yxogenium replied:
FlashPrint requires libudev.so.1, and the shared modules provide libudev.so.0. I asked for recommendation [in a reddit thread][1] and I was advised to use [eudev][2].

[1]: https://www.reddit.com/r/Fedora/comments/ams0xd/making_a_flatpak_for_flashprint_issue_with/
[2]: https://wiki.gentoo.org/wiki/Eudev

nedrichards commented:
bot, build com.flashforge.FlashPrint

nedrichards commented:
bot, build com.flashforge.FlashPrint

nedrichards commented:
I get the following error when installing:
>        sed: -e expression #1, char 48: unterminated `s' command
> flatpak install: Invalid extension for com.flashforge.FlashPrint in app com.flashforge.FlashPrint, removing.

scx commented on com.flashforge.FlashPrint.json:
```diff
@@ -0,0 +1,80 @@
+            "mv usr/* .",
+            "rmdir usr",
+            "mkdir -p export/share/applications",
+            "sed s/Icon=flashforge/Icon=com.flashforge.FlashPrint share/applications/FlashPrint.desktop > export/share/applications/com.flashforge.FlashPrint",
```
1. Invalid sed command. Should be:
```
sed 's/Icon=flashforge/Icon=com.flashforge.FlashPrint/'
```
or
```
sed 's/(Icon=).+/\1com.flashforge.FlashPrint/'
```
2. You can (and should) use `desktop-file-edit` or `desktop-file-install` instead of sed.
You can also use `rename-icon` and (if needed) `copy-icon`.

Yxogenium commented:
I removed the offending sed invocation: the desktop file is now installed and edited using `desktop-file-install`. Since `flashprint.deb` is only extracted at install time, I chose to bundle the `desktop-file-install` binary in the flatpak. For the same reason, I cannot use `rename-desktop` or `rename-icon`.

When I install FlashPrint from a local repo, I can now see an icon in my application launcher. Thank you again for all the pointers!

Yxogenium commented:
However, the icon fails to run the app. This is probably because by default the .desktop file contains
```
Exec=/usr/share/FlashPrint/FlashPrint
```
where it should be
```
Exec=/usr/bin/FlashPrint
```
(so that is runs the wrapper called `flashprint-bin` in this repo).

I tried to add the flags `--set-key=Exec --set-value=/usr/bin/FlashPrint` to the `desktop-file-install` invocation. It does not seem to solve the issue. Any ideas ?

Yxogenium commented:
It would probably be simpler to write our own `com.flashforge.FlashPrint.desktop` (as a plus we wouldn't need to bundle `desktop-file-install`). It just don't want to breach their license.

Yxogenium commented:
It would probably be simpler to write our own `com.flashforge.FlashPrint.desktop` (as a plus we wouldn't need to bundle `desktop-file-install`). I just don't want to breach their license.

nedrichards commented:
Right, rename-icon and copy-icon are only run at flatpak build time, not when the apply_extra script is run. So you'd need to take care of any renames in there. For icons we have a policy that including the app icon in the manifest itself is acceptable promotion of a program, so you can add that directly. (If you don't have an icon and appdata in the app before install it just won't show up in the app center, which means people won't find it).

nedrichards commented:
bot, build com.flashforge.FlashPrint

scx commented on com.flashforge.FlashPrint.json:
```diff
@@ -0,0 +1,85 @@
+            "mv usr/* .",
+            "rmdir usr",
+            "mv share/applications/FlashPrint.desktop share/applications/com.flashforge.FlashPrint.desktop",
+            "desktop-file-install --dir=export/share/applications --set-key=Exec --set-value=/app/bin/FlashPrint --set-icon=com.flashforge.FlashPrint share/applications/com.flashforge.FlashPrint.desktop",
```
Remove the comma (`,`) at the end.

scx commented:
bot, build com.flashforge.FlashPrint

frafra commented:
I am not able to install it as @flathubbot suggests (the url is not valid), nor to run it locally.

```
$ git clone --branch com.flashforge.FlashPrint --recursive https://github.com/Yxogenium/flathub.git
cd flathub
$ flatpak install --assumeyes org.kde.Sdk/x86_64/5.9 org.kde.Platform/x86_64/5.9
$ flatpak-builder build ./com.flashforge.FlashPrint.json --force-clean
$ flatpak-builder --run build/ com.flashforge.FlashPrint.json FlashPrint
```

Here's the output:
```
/app/bin/FlashPrint: line 3: /app/extra/share/FlashPrint/FlashPrint: No such file or directory
```

What about this fix?
```
sed -i 's;/app/extra/share/FlashPrint/FlashPrint;/app/bin/FlashPrint;' flashprint-bin
```

Now I do not have any errors, but the window would not just appear, I have an empty terminal and that's it. strace is looping over some missing files:

```
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
open("/app/lib/tls/x86_64/libtinfo.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/app/lib/tls/x86_64", 0x7ffe4d692390) = -1 ENOENT (No such file or directory)
open("/app/lib/tls/libtinfo.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/app/lib/tls", 0x7ffe4d692390)    = -1 ENOENT (No such file or directory)
open("/app/lib/x86_64/libtinfo.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/app/lib/x86_64", 0x7ffe4d692390) = -1 ENOENT (No such file or directory)
open("/app/lib/libtinfo.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/ffmpeg/lib/tls/x86_64/libtinfo.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/ffmpeg/lib/tls/x86_64", 0x7ffe4d692390) = -1 ENOENT (No such file or directory)
open("/usr/lib/ffmpeg/lib/tls/libtinfo.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/ffmpeg/lib/tls", 0x7ffe4d692390) = -1 ENOENT (No such file or directory)
open("/usr/lib/ffmpeg/lib/x86_64/libtinfo.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/ffmpeg/lib/x86_64", 0x7ffe4d692390) = -1 ENOENT (No such file or directory)
open("/usr/lib/ffmpeg/lib/libtinfo.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/app/lib/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/ffmpeg/lib/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/app/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/ffmpeg/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/gconv/gconv-modules.cache", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_MEASUREMENT", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_TELEPHONE", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_ADDRESS", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_NAME", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_PAPER", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_MESSAGES", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_MONETARY", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_COLLATE", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_TIME", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_NUMERIC", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/en_US.utf8/LC_CTYPE", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

Yxogenium commented:
Simply building the flatpak is not enough ; the executable (`/app/extra/share/FlashPrint/FlashPrint`) is downloaded at install time. I know it used to run correctly, but I can't test it right now:
```console
$ flatpak install --user https://dl.flathub.org/build-repo/883/com.flashforge.FlashPrint.flatpakref
**
flatpak:ERROR:app/flatpak-main.c:700:flatpak_run: assertion failed: (success || error)
Aborted (core dumped)
$
```

Yxogenium commented:
You can use this Makefile to run it locally:
```Makefile
.PHONY: build
build:
	flatpak-builder build/ ./com.flashforge.FlashPrint.json --force-clean

.PHONY: repo
repo:
	flatpak-builder --repo=repo --force-clean build com.flashforge.FlashPrint.json

.PHONY: install
install:
	flatpak --user remote-add --no-gpg-verify local-flashprint-repo repo
	flatpak --user install local-flashprint-repo com.flashforge.FlashPrint
```

And run `make build repo install`

frafra commented:
This is because the URL is not available: https://github.com/flatpak/flatpak/issues/2727

Your solution works, thank you! :-)

nedrichards commented:
bot, build com.flashforge.FlashPrint

(that will give a fresh testing repo)

nedrichards commented:
Repository has been created: https://github.com/flathub/com.flashforge.FlashPrint
